File: 1812/Domain/Deck.cs
using System;
using System.Collections.Generic;

namespace _1812.Domain
{
    public class Deck
    {
        private Stack<Card> deckCards;
        private List<Card> throwCards;
        private List<Card> allCards;
        public int DeckCount { get { return deckCards.Count; } }
        public int ThrowCount { get { return throwCards.Count; } }
        public int AllCount { get { return allCards.Count; } }
        private static List<Card> Shuffle(List<Card> toShuffle)
        {
            List<Card> data = toShuffle;
            for (int i = data.Count - 1; i >= 1; i--)
            {
                var rand = new Random();
                int j = rand.Next(i + 1);
                // обменять значения data[j] и data[i]
                Card temp = data[j];
                data[j] = data[i];
                data[i] = temp;
            }

            return data;
        }
        private List<Card> Take(int count)
        {
            List<Card> toReturn = new List<Card>();

            for (int i = 0; i < count; i++)
            {
                toReturn.Add(deckCards.Pop());
            }
            return toReturn;
        }
        public List<Card> GetCards(int count)
        {
            if (count <= deckCards.Count)
            {
                return Take(count);
            }
            else
            {
                List<Card> toReturn = Take(deckCards.Count);

                foreach(Card i in Shuffle(throwCards))
                {
                    deckCards.Push(i);
                }
                throwCards.Clear();

                toReturn.AddRange(Take(count - toReturn.Count));
                return toReturn;
            }
        }

        public void PutCards(List<Card> toPut)
        {
            throwCards.AddRange(toPut);
        }

        public Deck(List<Card> cards, List<Card> drop)
        {
            deckCards = new Stack<Card>();
            allCards = cards;
            foreach (Card i in Shuffle(cards))
            {
                deckCards.Push(i);
            }
            throwCards = drop;
        }

        public override string ToString()
        {
            return "Колода из "+AllCount.ToString()+" карт";
        }
    }
}

File: domain/deck.py
from typing import List
from domain.card import Card
import random

class Deck:

    def __init__(self, cards: List[Card], drop: List[Card]) -> None:
        self.all_cards = cards
        self.deck_cards = cards.copy()
        random.shuffle(self.deck_cards)
        self.throw_cards = drop

    def __str__(self) -> str:
        return 'Deck from ' + str(self.all_cards)

    def put_cards(self, to_put: List[Card]):
        self.throw_cards += to_put

    def _take(self, count: int) -> List[Card]:
        '''
        Removes `count` elements from deck and returns them
        '''
        return [self.deck_cards.pop(0) for _ in range(count)]

    def get_cards(self, count: int):
        if count <= len(self.deck_cards):
            return self._take(count)

        # shuffle to throw and move it to the main deck
        to_return = self._take(len(self.deck_cards))

        self.deck_cards = self.throw_cards
        random.shuffle(self.deck_cards)
        self.throw_cards = []

        return to_return + self._take(count - len(self.to_return))

    @property
    def deck_count(self):
        return len(self.deck_cards)

    @property
    def throw_count(self):
        return len(self.throw_cards)

    @property
    def all_count(self):
        return len(self.all_cards)
